WCAG 2.1 Development

jestonihpi opened the pull request:
R&D for WCAG 2.1 Implementation. The main goal was to follow the best practices as close as possible. 
## Description

<!-- Describe what you have changed or added. -->
Added aria tags.
<!-- Link to the support ticket(s) where appropriate. -->

## Testing instructions
<!-- Add instructions to help the reviewer test your code. -->
<!-- Include sample forms, add-ons or snippets where appropriate. -->

## Screenshots <!-- if applicable -->

## Checklist:
- [ ] I've tested the code.
- [ ] My code is easy to read, follow, and understand
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation / docblocks. <!-- Guidelines: http://docs.phpdoc.org/references/phpdoc/basic-syntax.html -->

## Additional Comments <!-- if applicable -->

debbieellis7 committed:
JS unit tests

debbieellis7 committed:
E2E tests

jestonisumande committed:
Initial commit. Remove aria-hidden on settings pages (ver1), add roles and attempts on svg icon checkboxes

jestonisumande committed:
Initial commit. Remove aria-hidden on settings pages (ver1), add roles and attempts on svg icon checkboxes

jakejackson1 commented on src/Helper/Helper_Abstract_Options.php:
```diff
@@ -1545,10 +1546,13 @@ public function multicheck_callback( $args ) {
-				echo '<input name="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" id="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" class="gfpdf_settings_' . $args['id'] . ' ' . $class . '" type="checkbox" value="' . $args['multi-key'] . '" ' . checked( $key, $enabled, false ) . ' ' . $required . ' />&nbsp;';
-				echo '<label for="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']">' . esc_html( $option ) . '</label>';
+				echo '<input name="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" id="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" class="gfpdf_settings_' . $args['id'] . ' ' . $class . '" type="checkbox" value="' . $args['multi-key'] . '" ' . checked( $key, $enabled, false ) . ' ' . $required . ' role="checkbox" ' . $aria_checked . ' />&nbsp;';
+				echo '<label for="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" aria-label="'. $aria_label .'">' . $label . '</label>';
```
All my reading suggests that `aria` tags aren't necessary on a `<label>` tag with a correctly referenced `for` target:

https://stackoverflow.com/questions/55814279/aria-label-vs-form-labels
https://www.w3.org/TR/WCAG20-TECHS/ARIA14.html
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute

Could you please provide reference martial that backs up why using both is appropriate?

jestonihpi replied:
As for my test Jake. aria-label is prioritized by screen reader. My goal was to design the labels to more user friendly in terms of audio. Thus $aria_label = str_replace( '_', ' ',$label ); e.g gravityforms_view_entries will be read as gravityforms view entries.
Anyways this item is still experimental.

jakejackson1 replied:
That makes sense, but we are talking about specific capability names [which are described in detail in the Gravity Forms documentation](https://docs.gravityforms.com/role-management-guide/). If they had read the GF docs, users would be looking for `gravityforms*underscore*view*underscore*entries`, rather than `gravityforms view entries`.

jakejackson1 commented on src/Helper/Helper_Abstract_Options.php:
```diff
@@ -1545,10 +1546,13 @@ public function multicheck_callback( $args ) {
 				$args['multi-option'] = $key;
 
 				$enabled = $this->get_form_value( $args );
+				$label = esc_html( $option ) ;
+				$aria_checked = !empty( $enabled ) ? 'aria-checked="true"' : 'aria-checked="false"';
+				$aria_label = str_replace( '_', ' ',$label );
```
This code needs to be linted

jakejackson1 commented on src/Helper/Helper_Abstract_Options.php:
```diff
@@ -1545,10 +1546,13 @@ public function multicheck_callback( $args ) {
 				echo '<div class="gfpdf-settings-multicheck-wrapper">';
-				echo '<input name="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" id="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" class="gfpdf_settings_' . $args['id'] . ' ' . $class . '" type="checkbox" value="' . $args['multi-key'] . '" ' . checked( $key, $enabled, false ) . ' ' . $required . ' />&nbsp;';
-				echo '<label for="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']">' . esc_html( $option ) . '</label>';
+				echo '<input name="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" id="gfpdf_settings[' . $args['id'] . '][' . $args['multi-key'] . ']" class="gfpdf_settings_' . $args['id'] . ' ' . $class . '" type="checkbox" value="' . $args['multi-key'] . '" ' . checked( $key, $enabled, false ) . ' ' . $required . ' role="checkbox" ' . $aria_checked . ' />&nbsp;';
```
This is an invalid use of `aria-checked`. See:

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/checkbox_role

jakejackson1 commented on src/Helper/Helper_Options_Fields.php:
```diff
@@ -237,7 +236,7 @@ public function get_registered_fields() {
 					'id'         => 'uninstaller',
 					'name'       => esc_html__( 'Uninstall Gravity PDF', 'gravity-forms-pdf-extended' ),
-					'desc'       => '<p class="alert error">' . esc_html__( 'This operation deletes ALL Gravity PDF settings and deactivates the plugin. If you continue, all settings, configuration, custom templates and fonts will be removed.', 'gravity-forms-pdf-extended' ) . '</p>',
+					'desc'       => '<p class="alert error" role="dialog">' . esc_html__( 'This operation deletes ALL Gravity PDF settings and deactivates the plugin. If you continue, all settings, configuration, custom templates and fonts will be removed.', 'gravity-forms-pdf-extended' ) . '</p>',
```
The use of `role="dialog"` isn't valid in this context:

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role

It's only used for dialog boxes, but this message displays right above the button it relates to.

jestonihpi replied:
This will be replaced with role="alert".

jakejackson1 replied:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role

This would suggest that the alert role should be used for validation messages and not for generate descriptions about an action. It seems that `aria-describedby` would be more practical here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute

jakejackson1 commented on src/Helper/Helper_PDF_List_Table.php:
```diff
@@ -228,6 +228,8 @@ public function column_cb( $item ) {
 			 data-fid="<?= $form_id ?>"
 			 src="<?= $this->gform->get_plugin_url() ?>/images/active<?= intval( $is_active ) ?>.svg"
 			 alt="<?= $label ?>"
+			 role="img"
+			 aria-label="<?= $label ?>"
```
`role="img"` definitely isn't the right approach here as it's assigned to a single image and not a container with a "group" of images https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Role_Img

. `arial-label` not correct either. Likely need aria-live on a new visible element: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions. Since GF has updated how the toggle looks this should be much easier to achieve.

jakejackson1 reviewed (changes requested):
Along with me notes, you need to rebase this branch on the latest `development` branch and then force push.

jestonisumande committed:
Fixes on invalid use of aria tags

jestonisumande committed:
Fixes on invalid use of aria tags

jestonisumande committed:
Apply aria-live with alert role on font installation message

jestonisumande committed:
Remove space

jestonisumande committed:
Update the PDF toggle state to match GF

jestonisumande committed:
Application of aria-label on gfpdf custom pdf list table

jestonisumande committed:
Remove roles and aria tags, based on wp's notification.

jakejackson1 committed:
Ensure WP Docker site uses locked JS dependancies

jestonisumande committed:
Apply form and listbox roles. Switch from div to ul on font listing

jestonisumande committed:
Apply form and listbox roles. Switch from div to ul on font listing

jakejackson1 commented on src/assets/js/react/components/FontManager/SearchBox.js:
```diff
@@ -120,17 +120,19 @@ export class SearchBox extends Component {
-        ref={node => (this.input = node)}
-        tabIndex='143'
-      />
+      <div role='form'>
```
Couple comments on this:

1. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Form_Role recommends using `<form>` where possible (but doesn't go into why you would choose to use `role="form"` instead of `<form>`). What research have you done into use the role-base alternative, over using the native `<form>` tag?

2. Regardless of which way we go, shouldn't we look at wrapping both the search and results in a FORM element, since the results include form-related actions (delete for example)?

jakejackson1 replied:
Also, it looks like `<form role="search">` is preferred https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Search_role

jakejackson1 committed:
Fix up the duplicate feature to work with the new status toggle. Also resolves jumping issue when running the duplicate command.

Fixes #1123

jakejackson1 committed:
Fix jQuery 3.5 deprecated warnings

jakejackson1 committed:
Remove unneeded JS

jakejackson1 committed:
Remove unneeded strings

jakejackson1 committed:
Resolve File Upload selector issue

Also fixes problem using field with GF No Conflict mode on

Fix #1135

jakejackson1 committed:
jQuery 3.5: Convert 'change' shorthand to long form

jakejackson1 committed:
Ensure Gravity PDF shortcode is automatically processed on Redirect page

Fixes #1134

jakejackson1 committed:
Fix PHP notice when no valid $form or $entry

This issue can occur if including the {:pdf:000000} merge tag prior to form submission.

Resolves #1146

jakejackson1 committed:
Only show active PDFs in entry export

Resolves #1136

jakejackson1 committed:
Inline the Font and Template Advanced buttons

Resolves #1124

jakejackson1 committed:
Fix PHPUnit error after limiting exported PDFs to active

jakejackson1 committed:
Resolve bugs related to deleting and validating fonts

Also tidied up unit-only tested files

debbieellis7 committed:
Remove kashida setting from UI

debbieellis7 committed:
Fix font manager selection bug

jakejackson1 committed:
Update composer managed packages

debbieellis7 committed:
Move selected font at the very top of the font list

jakejackson1 committed:
Add issue template for GitHub

jakejackson1 committed:
Revert move of LocalFilesystem and LocalFile classes

jakejackson1 committed:
Only allow font to be installed via template if not already exists

Resolves #1151

jakejackson1 committed:
Remove the legacy font code in a way that won't cause fatal errors

debbieellis7 committed:
Auto select newly installed font

jakejackson1 committed:
Update README for better formatting and remove unsupported versions of PHP

jakejackson1 committed:
Increase Docker PHP Memory Limit

jakejackson1 committed:
Exclude vendor_prefix from linter

jakejackson1 committed:
Exclude tmp folder from PHPCS

jakejackson1 committed:
Fix phpunit AJAX issues

debbieellis7 committed:
Fix font manager incorrectly set font bug

jestonisumande committed:
Updated role (log)

jestonisumande committed:
Added label on uninstaller form

jestonisumande committed:
Added a fix for enter key that disable form submission and triggers click

jakejackson1 committed:
Add GitHub Actions

jakejackson1 committed:
Update WP Tested Up To number

jakejackson1 committed:
When packaging, use the current checked out HEAD in GIT

jestonisumande committed:
Updates for https://app.clickup.com/t/51q3r7

jestonisumande committed:
Updates for https://app.clickup.com/t/51xt7w

jestonisumande committed:
Updates for https://app.clickup.com/t/51puuh

jestonisumande committed:
Updates for https://app.clickup.com/t/51q7qf Entries - Settings - PDF List

jestonisumande committed:
Updatesfor https://app.clickup.com/t/55xd75 Form - Settings - PDF / Settings - PDF  - Template Manager

jakejackson1 commented:
@jestonihpi please rebase your PR on the latest `development` branch and only include changes relevant to your work (i.e no code reformatting).

jestonisumande committed:
Updates for https://app.clickup.com/t/57tfjr - Issue with Multiple Banner

jestonisumande committed:
Added aria-live=polite on font listing

jakejackson1 commented:
Closed in favour of https://github.com/GravityPDF/gravity-pdf/pull/1177
